velox/functions/lib/aggregates/noisy_aggregation/sketch/SfmSketch.cpp

```diff
@@ -0,0 +1,370 @@
+  // Cap the number of trailing zeros to precision - 1, to avoid out of
+  // bound.
+  auto zeros =
+      std::min(precision_ - 1, numberOfTrailingZeros(hash, indexBitLength_));
+  setBitTrue(bucketIndex, zeros);
+}
+
+void SfmSketch::addIndexAndZeros(uint32_t bucketIndex, uint32_t zeros) {
+  auto numOfbuckets = numberOfBuckets(indexBitLength_);
+  VELOX_CHECK_GE(bucketIndex, 0, "Bucket index out of range.");
+  VELOX_CHECK_LT(bucketIndex, numOfbuckets, "Bucket index out of range.");
+  VELOX_CHECK_GE(zeros, 0, "Zeros must be GREATER than or equal to 0.");
+  VELOX_CHECK_LE(zeros, 64, "Zeros must be LESS than or equal to 64.");
+
+  // count of zeros in range [0, precision - 1]
+  zeros = std::min(precision_ - 1, zeros);
+  setBitTrue(bucketIndex, zeros);
+}
+
+uint32_t SfmSketch::computeIndex(uint64_t hash, uint32_t indexBitLength) {
+  validateIndexBitLength(indexBitLength);
+  constexpr uint32_t kBitWidth = sizeof(uint64_t) * 8;
+  return static_cast<uint32_t>(hash >> (kBitWidth - indexBitLength));
+}
+
+uint32_t SfmSketch::numberOfBuckets(uint32_t indexBitLength) {
+  validateIndexBitLength(indexBitLength);
+  return 1 << indexBitLength;
+}
+
+uint32_t SfmSketch::indexBitLength(uint32_t numberOfBuckets) {
+  VELOX_CHECK_GT(numberOfBuckets, 0, "numberOfBuckets must be greater than 0.");
+  VELOX_CHECK_EQ(
+      numberOfBuckets & (numberOfBuckets - 1),
+      0,
+      "numberOfBuckets must be power of 2.");
+  return static_cast<uint32_t>(std::log2(numberOfBuckets));
+}
+
+double SfmSketch::calculateRandomizedResponseProbability(double epsilon) {
+  if (epsilon == kNonPrivateEpsilon) {
+    return 0.0;
+  }
+  return 1.0 / (1.0 + exp(epsilon));
+}
+
+void SfmSketch::mergeWith(const SfmSketch& other) {
+  MersenneTwisterRandomizationStrategy randomizationStrategy;
+  mergeWith(other, &randomizationStrategy);
+}
+
+void SfmSketch::mergeWith(
+    const SfmSketch& other,
+    RandomizationStrategy* randomizationStrategy) {
+  VELOX_CHECK_EQ(
+      precision_,
+      other.precision_,
+      "Cannot merge two SFM sketches with different precision: {} vs. {}");
+  VELOX_CHECK_EQ(
+      indexBitLength_,
+      other.indexBitLength_,
+      "Cannot merge two SFM sketches with different indexBitLength: {} vs. {}");
+
+  uint64_t* target = reinterpret_cast<uint64_t*>(bits_.data());
+  const uint64_t* source =
+      reinterpret_cast<const uint64_t*>(other.bits_.data());
+
+  // If neither sketch is private, we just take the OR of the sketches.
+  if (!privacyEnabled() && !other.privacyEnabled()) {
+    bits::orBits(target, source, 0, static_cast<int32_t>(numberOfBits()));
+  } else {
+    // If either sketch is private, we combine using a randomized merge
+    double p1 = randomizedResponseProbability_;
+    double p2 = other.randomizedResponseProbability_;
+    double p = mergeRandomizedResponseProbabilities(p1, p2);
+    double normalizer = (1 - 2 * p) / ((1 - 2 * p1) * (1 - 2 * p2));
+
+    for (uint32_t i = 0; i < numberOfBits(); i++) {
```

Comment:
> numberOfBits()

Let's create a temp variable to store this value to avoid re-computing it multiple times.